sql/wsrep_trans_observer.h

```diff
@@ -265,7 +266,9 @@ static inline int wsrep_before_prepare(THD* thd, bool all)
   {
     DBUG_RETURN(ret);
   }
-  if ((ret= thd->wsrep_cs().before_prepare()) == 0)
+  wsrep::provider::seq_cb_t seq_cb{
+      thd, wsrep_parallel_slave_wakeup_subsequent_commits};
+  if ((ret= thd->wsrep_cs().before_prepare(&seq_cb)) == 0)
```

Comment:
@temeo Here we provide callback to Galera library but how the sequential consistency is maintained because this is called out-of-order? When the actual critical section is released, is it then on before_commit?

Reply:
The callback is effective only for parallel slave workers, see https://github.com/codership/mariadb-server/pull/459/files#diff-adadb13c7a06c1a6d6be490308b4cb130848c6e0b2ccca48dde12efeda10a498R3926. The workers call `wait_for_prior_commit()` from `thd->wsrep_parallel_slave_wait_for_prior_commit()` couple of lines above this, which makes workers' calls to `before_prepare()` serialized.